content/blog/2020-12-23-hof/index.md

```diff
@@ -0,0 +1,83 @@
+
+#### Code Breakdown
+
+I had three main functions: `free_vars`, which took in a function's arguments and body and determined the free variables it contained, `process_func`, which took in a function's arguments as well as its basic blocks and returned a list of extracted global functions as well as the original function's new block structure, and `process_q`, which took in a program's CFG and queued each of its functions for processing with `process_func`, adding on new global functions to extract other functions from to the queue. 
+
+#### `bril-txt` modification 
+
+I updated `briltxt.py` to translate Bril programs using HOFs in a textual representation into the JSON representation (and vice versa). This made it a lot easier to write and run test cases! 
+
+## What was hard to get right? 
+
+A saying (by Phil Karlton, allegedly) that gets riffed on a lot among computer scientists goes: *There are 2 hard problems in computer science: cache invalidation, naming things, and off-by-1 errors*. I encountered all of the above when implementing HOFs in Bril. 
+
+Function renaming was tricky as the mangling scheme I implemented made it hard to keep track of which function names mapped to each function in my cache of functions. I ended up scrapping the mangling scheme as it was never a problem in the example programs and I figured I could require unique names as part of the specification. I also had some issues with queueing extracted global functions into my lambda lifter; I was always missing the last one. I'm not sure if this was because I was working in OCaml where mutability is not welcomed, but after making the code more explicitly imperative this problem vanished. 
+
+I discovered that I couldn't take functions as arguments or return them since that would create a closure, which is a more general form of the nested functions I was implementing. To implement the desired functionality, I would need to develop a closure syntax, which was beyond what I was able to accomplish this semster. 
+
+## Was I successful? 
+
+I went through the Bril benchmarks and selected the ones that had multiple functions. I modified these programs to have multiple layers of nesting and/or multiple nested functions in a single function. These programs can be found in my repository under `test/hof/`. I then lifted these modified programs and compared the number of dynamic instructions I got to the number that interpreting the original programs directly yielded. In most cases, I had the same number of instructions, which is at least some indicator that I was doing something right methodologically. I noticed that for programs where I had multiple nested functions and/or multiple nested function calls, the number of instructions increased by 1 or 2. `binary-fmt.bril` was a notable exception to this: it had 10 extra instructions. I suspect that this was due to the recursive nature of one of the functions as well as the calls to another HOF inside the recursive body. 
```

Comment:
Perhaps some brief overview of what the programs were that you implemented would be useful here.

Reply:
Let me know if this is okay!

Reply:
Sounds good!